cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,504 @@
+            (this->value_squared == rhs.value_squared)
+        );
+    }
+};
+
+// --------------------------------------------------------------------------------------------------------
+/** -------------------------------------------------------------------------*
+ * @brief input struct with/without null bitmask
+ * A helper struct struct for `column_input_iterator`
+ * `ColumnInput.at(gdf_index_type id)` computes
+ * `data` value and valid flag at `id` and construct and return
+ * `T_output(data, valid flag)`.
+ * If nulls_present = true, the valid flag is always true.
+ * If nulls_present = false, the valid flag corresponds to null bitmask flag
+ * at `id`, and the data value = (is_valid(id))? data[id] : identity;
+ *
+ * @tparam  T_output a scalar data type to be output
+ * @tparam  T_element a native element type (cudf dta type) of input element array
+ *           and `identity` value which is used when null bitmaps flag is false.
+ * @tparam  nulls_present if true, this struct holds only data array.
+ *            else, this struct holds data array and bitmask array and identity value
+ * -------------------------------------------------------------------------**/
+template<typename T_output, typename T_element, bool nulls_present=true>
+struct ColumnInput;
+
+// @overload ColumnInput<T_output, T_element, false>
+template<typename T_output, typename T_element>
+struct ColumnInput<T_output, T_element, false>{
+    const T_element *data;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnInput(const T_element *_data, const bit_mask::bit_mask_t *_valid=nullptr, T_element _identity=T_element{0})
+    : data(_data){};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnInput(const T_element *_data, const gdf_valid_type*_valid, T_element _identity)
+    : ColumnInput(_data, reinterpret_cast<const bit_mask::bit_mask_t*>(_valid), _identity) {};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_output at(gdf_index_type id) const {
+        return T_output(data[id], true);
+    };
+};
+
+// @overload ColumnInput<T_output, T_element, true>
+template<typename T_output, typename T_element>
+struct ColumnInput<T_output, T_element, true>{
+    const T_element *data;
+    const bit_mask::bit_mask_t *valid;
+    T_element identity;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnInput(const T_element *_data, const bit_mask::bit_mask_t *_valid, T_element _identity)
+    : data(_data), valid(_valid), identity(_identity){
+#if  !defined(__CUDA_ARCH__)
+        // verify valid is non-null, otherwise, is_valid() will crash
+        CUDF_EXPECTS(valid != nullptr, "non-null bit mask is required");
+#endif
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnInput(const T_element *_data, const gdf_valid_type*_valid, T_element _identity)
+    : ColumnInput(_data, reinterpret_cast<const bit_mask::bit_mask_t*>(_valid), _identity) {};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    T_output at(gdf_index_type id) const {
+        return T_output(get_value(id), is_valid(id));
```

Comment:
This is computing `is_valid` twice here. How about just get rid of the `get_value` and `is_valid` private members and do this: 

```
    T_output at(gdf_index_type id) const {
        return is_valid(valid, id) ? T_output(data[id], true) : T_output(identity, false); 
    }
```

Reply:
Fixed.

Reply:
It looks like this hasn't actually been changed yet.